db/snapshot/backup_task.cc

```diff
@@ -103,52 +105,89 @@ future<> backup_task_impl::upload_component(shared_ptr<s3::client>& client, sstr
-    auto snapshot_dir_lister = directory_lister(_snapshot_dir, lister::dir_entry_types::of<directory_entry_type::regular>());
-    auto cln = _snap_ctl.storage_manager().container().local().get_endpoint_client(_endpoint);
+    auto snapshot_dir_lister = directory_lister(snapshot_dir, lister::dir_entry_types::of<directory_entry_type::regular>());
+    std::vector<std::tuple<size_t, sstring>> backup_files;
+    std::optional<directory_entry> component_ent;
 
-    for (;;) {
-        std::optional<directory_entry> component_ent;
+    do {
         try {
             component_ent = co_await snapshot_dir_lister.get();
-        } catch (...) {
-            if (!ex) {
-                ex = std::current_exception();
-                break;
+            if (component_ent) {
+                auto component_name = snapshot_dir / component_ent->name;
+                auto size = co_await file_size(component_name.native());
+                backup_files.emplace_back(size, std::move(component_ent->name));
             }
-        }
-        if (!component_ent.has_value()) {
+        } catch (...) {
+            ex = std::current_exception();
             break;
         }
-        auto gh = uploads.hold();
+    } while (component_ent);
+    co_await snapshot_dir_lister.close();
+    if (ex)
+        co_await coroutine::return_exception_ptr(std::move(ex));
 
-        // Pre-upload break point. For testing abort in actual s3 client usage.
-        co_await utils::get_local_injector().inject("backup_task_pre_upload", utils::wait_for_message(std::chrono::minutes(2)));
+    std::ranges::sort(backup_files, std::greater());
+    co_return backup_files;
+}
+} // namespace
 
-        std::ignore = upload_component(cln, component_ent->name).handle_exception([&ex] (std::exception_ptr e) {
-            // keep the first exception
-            if (!ex) {
-                ex = std::move(e);
-            }
-        }).finally([gh = std::move(gh)] {});
-        co_await coroutine::maybe_yield();
-        co_await utils::get_local_injector().inject("backup_task_pause", utils::wait_for_message(std::chrono::minutes(2)));
-        if (impl::_as.abort_requested()) {
-            ex = impl::_as.abort_requested_exception_ptr();
-            break;
-        }
+future<> backup_task_impl::do_backup() {
+    if (!co_await file_exists(_snapshot_dir.native())) {
+        throw std::invalid_argument(fmt::format("snapshot does not exist at {}", _snapshot_dir.native()));
     }
 
-    co_await snapshot_dir_lister.close();
-    co_await uploads.close();
-    if (ex) {
-        co_await coroutine::return_exception_ptr(std::move(ex));
+    auto backup_files = co_await get_backup_files(_snapshot_dir);
+    std::atomic_size_t counter = 0;
+
+    sharded<abort_source> as;
+    co_await as.start();
+    gate as_gate;
+    auto s = _as.subscribe([&]() noexcept {
+        auto h = as_gate.hold();
+        std::ignore = smp::invoke_on_all([&as, ex = _as.abort_requested_exception_ptr()] {
```

Comment:
This is racy. In case `_as` is aborted _after_ the `co_await as.stop()` is called below, the invoke-on-all will be called on stopped sharded<abort_source>

Reply:
right
two options
1) I can add artificial scope, so the `subscription` object is destroyed before `as.stop()`
2) switch back to `std::vector<abort_source>` the way it is done in `sstables_loader`
WDYS?